pubsub/consumer.go

```diff
@@ -0,0 +1,198 @@
+			log.Error("Updating heardbeat", "consumer", c.id, "error", err)
+		}
+		select {
+		case <-ctx.Done():
+			log.Error("Error keeping alive", "error", ctx.Err())
+			return
+		case <-time.After(KeepAliveInterval):
+		}
+	}
+}
+
+// Consumer first checks it there exists pending message that is claimed by
+// unresponsive consumer, if not then reads from the stream.
+func (c *Consumer) Consume(ctx context.Context) (*Message, error) {
+	log.Debug("Attempting to consume a message", "consumer-id", c.id)
+	msg, err := c.checkPending(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("consumer: %v checking pending messages with unavailable consumer: %w", c.id, err)
+	}
+	if msg != nil {
+		return msg, nil
+	}
+	res, err := c.client.XReadGroup(ctx, &redis.XReadGroupArgs{
+		Group:    c.groupName,
+		Consumer: c.id,
+		// Receive only messages that were never delivered to any other consumer,
+		// that is, only new messages.
+		Streams: []string{c.streamName, ">"},
+		Count:   1,
+		Block:   time.Millisecond, // 0 seems to block the read instead of immediately returning
+	}).Result()
+	if errors.Is(err, redis.Nil) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("reading message for consumer: %q: %w", c.id, err)
+	}
+	if len(res) != 1 || len(res[0].Messages) != 1 {
+		return nil, fmt.Errorf("redis returned entries: %+v, for querying single message", res)
+	}
+	log.Debug(fmt.Sprintf("Consumer: %s consuming message: %s", c.id, res[0].Messages[0].ID))
+	return &Message{
+		ID:    res[0].Messages[0].ID,
+		Value: res[0].Messages[0].Values[msgKey],
+	}, nil
+}
+
+func (c *Consumer) ACK(ctx context.Context, messageID string) error {
+	log.Info("ACKing message", "consumer-id", c.id, "message-sid", messageID)
+	_, err := c.client.XAck(ctx, c.streamName, c.groupName, messageID).Result()
+	return err
+}
+
+// Check if a consumer is with specified ID is alive.
+func (c *Consumer) isConsumerAlive(ctx context.Context, consumerID string) bool {
+	val, err := c.client.Get(ctx, keepAliveKey(consumerID)).Int64()
+	if err != nil {
+		return false
+	}
+	return time.Now().UnixMilli()-val < 2*int64(KeepAliveTimeout.Milliseconds())
+}
+
+func (c *Consumer) lockPending(ctx context.Context, consumerID string) bool {
+	acquired, err := c.client.SetNX(ctx, pendingMessagesKey, consumerID, KeepAliveInterval).Result()
+	if err != nil || !acquired {
+		return false
+	}
+	return true
+}
+
+func (c *Consumer) unlockPending(ctx context.Context) {
+	log.Debug("Releasing lock", "consumer-id", c.id)
+	c.client.Del(ctx, pendingMessagesKey)
+
+}
+
+// checkPending lists pending messages, and checks unavailable consumers that
+// have ownership on pending message.
+// If such message and consumer exists, it claims ownership on it.
+func (c *Consumer) checkPending(ctx context.Context) (*Message, error) {
+	// Locking pending list avoid the race where two instances query pending
```

Comment:
I liked the idea of checkPending from consumers, but this lock makes me question it. If a consumer dies holding the lock there is no way to recover the producer.
In our system - there could be many consumers, and they could be brought up/down unexpectedly, but nly one producer. If the producer (nitro node) dies recovery will not be immediate.
That's why I think we can let the producer check for consumer livelivess, and log + resend tasks that were scheduled to a stuck/dead consumer.

Reply:
if you look at PromiseInterface - I like the idea of having the producer's Produce return something that implements the same interface which will return the execution result when done, and during Await it could poll for consumer liveness and either fail or resend

Reply:
Move the retry logic to the producer as discussed. Will implement promiseInterface approach in the child pr.